internal/datastore/spanner/caveat.go

```diff
@@ -3,23 +3,122 @@ package spanner
 import (
 	"context"
 	"fmt"
+	"time"
+
+	"cloud.google.com/go/spanner"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc/codes"
 
 	"github.com/authzed/spicedb/pkg/datastore"
 	core "github.com/authzed/spicedb/pkg/proto/core/v1"
 )
 
 func (sr spannerReader) ReadCaveatByName(ctx context.Context, name string) (*core.CaveatDefinition, datastore.Revision, error) {
-	return nil, datastore.NoRevision, fmt.Errorf("unimplemented caveat support in datastore")
+	ctx, span := tracer.Start(ctx, "ReadCaveatByName", trace.WithAttributes(attribute.String("name", name)))
+	defer span.End()
+
+	caveatKey := spanner.Key{name}
+	row, err := sr.txSource().ReadRow(ctx, tableCaveat, caveatKey, []string{colCaveatDefinition, colCaveatTS})
+	if err != nil {
+		if spanner.ErrCode(err) == codes.NotFound {
+			return nil, datastore.NoRevision, datastore.NewCaveatNameNotFoundErr(name)
+		}
+		return nil, datastore.NoRevision, fmt.Errorf(errUnableToReadCaveat, err)
+	}
+	var serialized []byte
+	var updated time.Time
+	if err := row.Columns(&serialized, &updated); err != nil {
+		return nil, datastore.NoRevision, fmt.Errorf(errUnableToReadCaveat, err)
+	}
+
+	loaded := &core.CaveatDefinition{}
+	if err := loaded.UnmarshalVT(serialized); err != nil {
+		return nil, datastore.NoRevision, err
+	}
+	return loaded, revisionFromTimestamp(updated), nil
 }
 
 func (sr spannerReader) ListCaveats(ctx context.Context, caveatNames ...string) ([]*core.CaveatDefinition, error) {
-	return nil, fmt.Errorf("unimplemented caveat support in datastore")
+	ctx, span := tracer.Start(ctx, "ListCaveats", trace.WithAttributes(attribute.StringSlice("names", caveatNames)))
+	defer span.End()
+
+	keyset := spanner.AllKeys()
+	if len(caveatNames) > 0 {
+		keys := make([]spanner.Key, 0, len(caveatNames))
+		for _, n := range caveatNames {
+			keys = append(keys, spanner.Key{n})
+		}
+		keyset = spanner.KeySetFromKeys(keys...)
+	}
+	iter := sr.txSource().Read(
```

Comment:
Is Spanner ordering by any key implicitly? Would it be possible for the output to have a non-deterministic order?

Reply:
Is order implied in this interface though?

Reply:
I didn't think order mattered here?

Reply:
Well, I guess it makes API consumption easier - I think I added it in the first place for some tests, I'd need to revisit if that requirement still holds. CRDB/PSQL implementation do certainly have the order by clause.